lib/mutant/expression/method.rb

```diff
@@ -3,44 +3,44 @@ class Expression
 
     # Explicit method expression
     class Method < self
+      include Anima.new(:scope_name, :scope_symbol, :method_name)
+      private(*anima.attribute_names)
 
       MATCHERS = IceNine.deep_freeze(
         '.' => Matcher::Methods::Singleton,
         '#' => Matcher::Methods::Instance
       )
+      private_constant(*constants(false))
 
-      register(
-        /\A(?<scope_name>#{SCOPE_PATTERN})(?<scope_symbol>[.#])(?<method_name>#{METHOD_NAME_PATTERN})\z/
-      )
+      REGEXP = /\A(?<scope_name>#{SCOPE_PATTERN})(?<scope_symbol>[.#])(?<method_name>#{METHOD_NAME_PATTERN})\z/.freeze
+
+      # Return syntax
+      #
+      # @return [String]
+      #
+      # @api private
+      #
+      def syntax
+        [scope_name, scope_symbol, method_name].join
+      end
+      memoize :syntax
 
       # Return method matcher
       #
       # @param [Env] env
       #
-      # @return [Matcher::Method]
+      # @return [Matcher]
       #
       # @api private
       #
       def matcher(env)
         methods_matcher = MATCHERS.fetch(scope_symbol).new(env, scope)
-        method = methods_matcher.methods.detect do |meth|
-          meth.name.equal?(method_name)
-        end or fail NameError, "Cannot find method #{method_name}"
-        methods_matcher.matcher.build(env, scope, method)
+
+        Matcher::Filter.new(methods_matcher, ->(subject) { subject.expression.eql?(self) })
```

Comment:
> I think that only "DSL" methods should capture blocks, and everything else is better of with an simple traditional explicit argument.

I don't necessarily agree with this. I think if the language provides the construct, provided there's nothing inherently wrong with it (unlike some JS "features"), I think it's better to use a more idiomatic approach when possible.

The block mechanism in ruby is well understood, and it's probably not idiomatic to create a proc and pass it in as the last argument in a method call. The only reason I might consider passing it in as an explicit argument might be if it's a richer object that supports `#call`, and I don't want to reduce it to something more primitive like a `Proc`.